components/common/src/liveliness_checker.rs

```diff
@@ -0,0 +1,467 @@
+                               reason)
+    }
+
+    /// In general, the return of mark_thread_alive must be used, to help ensure that threads
+    /// are unregistered from checking when they exit their work loop. However, this does not
+    /// apply to divergent threads (that is, ones that never return), so instead of storing
+    /// an ignored value to satisfy the `must_use` of `mark_thread_alive`, call this function
+    /// to indicate the intent to never terminate.
+    ///
+    /// See http://doc.rust-lang.org/std/primitive.never.html
+    ///
+    /// ```
+    /// use habitat_common::liveliness_checker;
+    ///
+    /// fn run_loop() -> ! {
+    ///     loop {
+    ///         liveliness_checker::mark_thread_alive().and_divergent();
+    ///     }
+    /// }
+    /// ```
+    pub fn and_divergent(self) {}
+}
+
+fn unregister_thread_impl<T, E: ToString>(_: CheckedThread,
+                                          statuses: &mut ThreadStatusMap,
+                                          reason: Result<T, E>)
+                                          -> ThreadUnregistered<T, E> {
+    let thread = thread::current();
+    let thread_id = &thread.id();
+
+    match &reason {
+        Ok(_) => {
+            statuses.remove(thread_id);
+        }
+        Err(e) => {
+            if let Some(entry) = statuses.get_mut(thread_id) {
+                entry.1 = Status::DeadWithError { time_of_death: Instant::now(),
+                                                  error:         e.to_string(), };
+            } else {
+                error!("unregister_thread called for untracked thread: {} id: {:?}",
+                       thread.name().unwrap_or("Unnamed thread"),
+                       thread_id);
+            }
+        }
+    }
+
+    ThreadUnregistered(reason)
+}
+
+/// Call once per binary to start the thread which will check that all the threads that
+/// call `mark_thread_alive` continue to do so.
+pub fn spawn_thread_alive_checker() {
+    thread::Builder::new().name("thread-alive-check".to_string())
+                          .spawn(|| -> ! {
+                              let delay = ThreadAliveCheckDelay::configured_value().into();
+                              let threshold = ThreadAliveThreshold::configured_value().into();
+                              let max_time_since_death =
+                                  ThreadDeadIgnoreDelay::configured_value().into();
+                              loop {
+                                  let statuses =
+                                      &mut THREAD_STATUSES.lock()
+                                                          .expect("THREAD_STATUSES poisoned");
+                                  check_thread_heartbeats(statuses, threshold);
+                                  log_dead_threads(statuses);
+                                  cull_dead_threads(statuses, max_time_since_death);
+                                  thread::sleep(delay);
```

Comment:
Wait, how will we know _this_ thread is still alive?

:trollface: _Quis custodiet ipsos custodes?_ :trollface:

Reply:
I nearly added it, before considering how pointless that would be